### ISSUE TITLE
Extensible blocks

### ISSUE THREAD

**Issue by mwort** [coderedcorp/coderedcms#305]
#### Is your feature request related to a problem?
The blocks provided by coderedcms are great but they are difficult to extend consistently. E.g. adding a `code block` to the `CONTENT_STREAMBLOCKS` is currently only possible by overriding the `body` in all of the page models where this is required.

#### Describe the solution you'd like
#43 hinted on a solution that has so far not been implemented (correct me if I'm wrong). It would probably require all `*_STREAMBLOCK` globals in `blocks/__init__.py` to be converted to functions that will assemble the list of blocks depending on settings during the model construction. The setting could be a dictionary with entries pointing to a module with global list variables, e.g. something like this:
```
ADDITIONAL_BLOCKS = 'website.blocks'  # eg. contains CONTENT_STREAMBLOCKS
```

#### Describe alternatives you've considered
See example with `body` above.

#### Additional context
I can give this a go and push a PR if you think this would be the right way to do this.

**Comment by vsalvino** [comment]
I agree - we tried to make these blocks more dynamic at first, but it creates a lot of problems with migrations, because any change to a streamfield must trigger a django migration.

I would be interested to see what your solution looks like. Can't guarantee it will be merged but I would definitely give it a thorough review. Up until now we have not come up with a good way to easily add more blocks other than overriding the `body` field per model, or adding creating an abstract model and then inheriting from that everywhere.

**Comment by ayushin** [comment]
@mwort I'd like to see that PR :-)

**Comment by ayushin** [comment]
@vsalvino I completely understand the concern, however it is better to give that flexibility to a developer (perhaps with a warning) and leave the choice up to them if they know what they are doing.

I imagine a good approach would be to have `CODEREDCMS_EXTRA_CONTENT_BLOCKS` or similar in settings.

**Comment by ayushin** [comment]
Or something like this?
```
def register_layout_streamblock(name):
    def inner(block):
        if name not in get_layout_streamblocks():
            get_layout_streamblocks().append((name, block(CONTENT_STREAMBLOCKS)))
        return block
    return inner
```

**Comment by vsalvino** [comment]
Yes - what you have is exactly right, however the problem is that it breaks Django migrations (it will created endless migrations) due to how the wagtail streamfield blocks are actually part of the database migrations - so they cannot be dynamic. This is the number 1 issue we have struggled with in trying to build coderedcms to be dynamic in any fashion.

So further research into the Wagtail streamfield, and Django migrations is required.

**Comment by ayushin** [comment]
So I am hitting this issue very hard three-fold:

1. Adding own blocks - can be worked around by creating re-using existing blocks fields (i.e. subtitle with |safe for some extra html field) and using custom extra templates

2. Integrating `https://pypi.org/project/wagtailstreamforms/` (perhaps an idea anyway to have that in coderedcms by default?)

3. And the worst by adding an extra language with `wagtail-modeltranslation` - as it creates migrations in both `coderedcms` and `wagtailstreamforms` every time a new language is added to `LANGUAGES`

@vsalvino 

So far the cleanest solution I see is keeping the abstract models in `coderedcms` and moving all the managed models into `project_template` together with the settings such as `HTML_STREAMBLOCKS` from `blocks/__init__.py`

This would keep the blocks definitions inside of the `coderedcms` but all the migrations will be in the project and would allow clean and controlled upgrages.

**Comment by vsalvino** [comment]
Regarding your list:

1. To add your own blocks, override the "body" field on your models. You can build your own streamfield from our various blocks as defined in `coderedcms/blocks/__init__.py`, and mix in your own custom blocks as well.

2. Stream forms are already integrated: https://docs.coderedcorp.com/cms/stable/features/page_types/stream_forms.html

3. What you are describing is an issue which can be temporarily worked around by being "careful" with migrations - but this is by far our biggest and hardest bug right now. See
* https://github.com/coderedcorp/coderedcms/issues/56
* https://github.com/coderedcorp/coderedcms/issues/44

**Comment by ayushin** [comment]
1. Yes, but I would like to to insert my blocks inside of bootstrap column (two levels down from the body) - do I miss something?

2. Talking about `https://github.com/labd/wagtailstreamforms` - you can add these forms to any page inside of any streamfield, is it same?

3. Yes....

**Comment by vsalvino** [comment]
For custom blocks, you'd have to define the entire streamfield as so (see 'my custom block'). And then set that as the `body` field of your model.

```python
LAYOUT_STREAMBLOCKS = [
    ('hero', HeroBlock([
        ('row', GridBlock(CONTENT_STREAMBLOCKS + [('my custom block', MyBlock())])),
        ('cardgrid', CardGridBlock([
            ('card', CardBlock()),
        ])),
        ('html', blocks.RawHTMLBlock(icon='code', classname='monospace', label=_('HTML'))),
    ])),
    ('row', GridBlock(CONTENT_STREAMBLOCKS)),
    ('cardgrid', CardGridBlock([
        ('card', CardBlock()),
    ])),
    ('html', blocks.RawHTMLBlock(icon='code', classname='monospace', label=_('HTML'))),
]
```

**Comment by ayushin** [comment]
This worked. Combined with `./manage.py sync_translation_fields` I'm now finally back to the mainstream coderedcms package!

**Comment by ayushin** [comment]
OK, so what's the latest recommendation to add custom STREAMBLOCKS to non-pages, like snippets and content walls?

I'm using https://github.com/labd/wagtailstreamforms with great results and would like to embed them into content walls too.

@vsalvino Is it (wagtailstreamforms) something to replace the form pages with?

**Comment by ayushin** [comment]
I am trying to inherit ContentWall with ContentPopup(ContentWall) and then override content_walls on all my pages, but I got stuck with this:

```
django.core.exceptions.FieldError: Local field 'content' in class 'ContentPopup' clashes with field of the same name from base class 'ContentWall'.
```

Any ideas of another solution to be able to add custom streamfield blocks to content walls? 🙇‍♂️

**Comment by ayushin** [comment]
OK, so revisiting this, will open a new issue for wagtailstreamforms, I've been banging my head against the wall for quite some time but any mechanism with EXTRA_*_STREAMFIELD_BLOCKS would lead to unmanageable migrations in each coderedcms installation and breakage on subsequent upgrade.

I'm thinking the solution could be moving al the CODEREDCMS streamfield migrations to a separate app or website, haven't given it a deeper think yet, just occurred to me as I type it.

I'm really struggling with it because I'd like to use wagtailstreamforms in the snippets (reusable content and content walls) and there's no good solution for it.

https://github.com/coderedcorp/coderedcms/issues/368

**Comment by vsalvino** [comment]
If you are trying to add a form to content wall... I would recommend creating the form page separately, then use the page preview block within the content wall, pointing to your new form page.

Yes, you are right about adding custom block types, it does create migration problems. The solution is that coderedcms needs to refactor its concrete models containing a streamfield to be abstract models... this is a design flaw that will unfortunately require a major update to sort out (#56)

**Comment by awhileback** [comment]
Hello everyone! Thanks for the great boilerplate for Wagtail, vsalvino and crew.

> I'm thinking the solution could be moving al the CODEREDCMS streamfield migrations to a separate app or website, haven't given it a deeper think yet, just occurred to me as I type it.

That's precisely what I did when working through similar issues (mainly with integrating third party blocks), I could't think of a better way. Blocks without formatting can just be included (such as alternate editors, markdown, etc) but when I wanted to embed uploaded media, as mentioned in #15, the only way I can think of to include such a block with the `advsettings` class attached is to create a new StructBlock from the included BaseBlock that includes `advsettings`, and then add the third party media block to the StructBlock.

Then there's the matter of how to get attributes from the third party block's model into the StructBlock, which I presume having a javascript listener to populate the fields if the media block selection changes is the proper way for, or, alternatively, allowing all of the fields from that model to be "blank" and then overriding the save method to include everything that was specified in the parent block's model.

In any case all of this will require a migration, so breaking the CMS out into an app seemed like the most serviceable way to do so.

A lot of this seems to be inherent limitations of wagtail. The structblock is provided to combine other blocks, but there doesn't seem to be a standardized way to manage the data in/from child blocks that provide required data without user input.

**Comment by minholi** [comment]
I tried to figure a way to do this and what I did was fork de repo, create a custom block and add coderedcms as a module into my website project repository. Maybe not the best way, but it works.

**Comment by vsalvino** [comment]
Best solution for now would be to create an official doc on how to make your own streamfield (see my comment above).
